spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/KafkaBackoffAwareMessageListenerAdapter.java

```diff
@@ -69,19 +69,18 @@ public void onMessage(ConsumerRecord<K, V> data) {
 	@Override
 	public void onMessage(ConsumerRecord<K, V> data, Acknowledgment ack, Consumer<?, ?> consumer) throws KafkaBackoffException {
 		maybeGetBackoffTimestamp(data)
-				.ifPresent(timestamp -> this.kafkaConsumerBackoffManager.maybeBackoff(createContext(data, timestamp)));
+				.ifPresent(nextExecutionTimestamp -> this.kafkaConsumerBackoffManager.maybeBackoff(createContext(data, nextExecutionTimestamp)));
 		super.getDelegate().onMessage(data, ack, consumer);
 		ack.acknowledge();
```

Comment:
@tomazfernandes I just noticed this - why are we acking here? Manual acks are the the responsibility of the user listener.

I didn't notice that you hard-wire the ack mode...

```
containerFactory.getContainerProperties().setAckMode(ContainerProperties.AckMode.MANUAL_IMMEDIATE);
```

In the configurer. What is the rationale here?

Thanks.

Reply:
Hi @garyrussell, thanks for reaching out.

I needed to make sure the app committed the message's offset after a processing goes wrong and I publish it to another queue, so that there's no chance of processing the original message again. 

In my testing in the early days of this PR I remember having to set that AckMode.MANUAL_IMMEDIATE to be able to guarantee this happening properly, but maybe I missed out on something back then.

But now that you are calling that out, would that acking still work if the user sets a different AckMode and having only ackAfterHandle set to true in the error handler?

If so we might be able to remove this hardcoding and also the exception from the other onMessage method.

Thanks

Reply:
Also, I'm remembering better now, I think back then I might have had a wrong understanding that the offset needed to be committed on the broker for the message not to be consumed again, when actually the consumer can handle it locally before the actual ack goes to the server.

Reply:
Does that make sense? Also, is it something you’ll look into or do you prefer that I do? I’m fine with both.

Reply:
I have to go to a funeral tomorrow (Friday), so I won't be here.

I was more concerned about a possible NPE, but your hard-coding avoids that (I was working on configuring `@Nullable` for Kotlin compatibility and noticed that the ack must not be null - see the latest commit on master).

I think we should allow the ackmode to be configured, but recommend the use of RECORD in this scenario.

Furthermore, I think we need to modify the DLPR to (at least optionally) wait on the send `Future` to make sure it was successful before acking the delivery (that's just a general comment, not related to this PR, but since we instantiate a DLPR in the DLPRF, we would need to set it somehow.

If you have time to explore this while I am away, that would be great.